src/dotnet/Chat.Service/Chats.cs

```diff
@@ -304,6 +308,140 @@ async Task Remove(ChatEntryId entryId1)
+            clonedAuthors.Add(clonedAuthor);
+        }
+        var authorMap = templateAuthors
+            .Where(a => a != null)
+            .Join(clonedAuthors,
+                a => a!.UserId,
+                a => a.UserId,
+                (l, r) => (TemplateAuthorId: l!.Id, CloneAuthorId: r.Id))
+            .ToDictionary(x => x.TemplateAuthorId, x => x.CloneAuthorId);
+        var roleAuthors = authorRoles
+            .SelectMany(x => x.Roles, (x, r) => (x.AuthorId, Role: r))
+            .Where(x => x.Role.SystemRole is not SystemRole.Anyone and not SystemRole.None and not SystemRole.Owner) // Owner is already registered
+            .GroupBy(x => x.Role.Id,
+                (_, xs) => {
+                    var tuples = xs.ToList();
+                    return (tuples.FirstOrDefault().Role, AuthorIds: tuples.Select(x => authorMap[x.AuthorId]).ToImmutableArray());
+                })
+            .ToList();
+
+        foreach (var (role, roleAuthorIds) in roleAuthors) {
+            var createOwnersRoleCmd = new IRolesBackend.ChangeCommand(cloned.Id, default, null, new() {
+                Create = new RoleDiff {
+                    Picture = role.Picture,
+                    Name = role.Name,
+                    SystemRole = role.SystemRole,
+                    Permissions = role.Permissions,
+                    AuthorIds = new SetDiff<ImmutableArray<AuthorId>, AuthorId> {
+                        AddedItems = roleAuthorIds,
+                    },
+                },
+            });
+            await Commander.Call(createOwnersRoleCmd, cancellationToken).ConfigureAwait(false);
+        }
+
+        // join guest author
+        var avatarIds = await Avatars.ListOwnAvatarIds(session, cancellationToken).ConfigureAwait(false);
+        var avatars = await avatarIds
+            .Select(aId => Avatars.GetOwn(session, aId, cancellationToken))
+            .Collect().ConfigureAwait(false);
+        var guestAvatar = avatars
+            .Where(a => a != null)
+            .FirstOrDefault(a => OrdinalEquals(a!.Name, Avatar.GuestName));
+        if (guestAvatar == null) {
+            var createAvatarCommand = new IAvatars.ChangeCommand(session, Symbol.Empty, null, new Change<AvatarFull> {
+                Create = new AvatarFull {
+                    UserId = account.Id,
+                    Name = "Guest",
+                }.WithMissingPropertiesFrom(account.Avatar),
+            });
+            var newAvatar = await Commander.Call(createAvatarCommand, cancellationToken).ConfigureAwait(false);
+            guestAvatar = newAvatar;
+        }
+        var createAuthorCommand = new IAuthorsBackend.UpsertCommand(
+            chatId,
+            AuthorId.None,
+            account.Id,
+            ExpectedVersion: null,
+            new AuthorDiff {
+                AvatarId = guestAvatar.Id,
+            });
+        await Commander.Run(createAuthorCommand, cancellationToken).ConfigureAwait(false);
```

Comment:
Is it possible to replace 'join guest author' block with IAuthors.JoinCommand?

Reply:
It's not safe to allow join to anyone, so we use backend command to add guest author explicitly to the new chat (create for this guest user)